executor/src/witgen/jit/identity_queue.rs

```diff
@@ -23,53 +23,115 @@ use super::{prover_function_heuristics::ProverFunction, variable::Variable};
-    queue: BTreeSet<QueueItem<'a, T>>,
-    occurrences: Rc<HashMap<Variable, Vec<QueueItem<'a, T>>>>,
+pub struct IdentityQueue<'ast, 'queue, T: FieldElement> {
+    items: &'queue Vec<QueueItem<'ast, T>>,
+    in_queue: Vec<bool>,
+    identity_queue: VecDeque<usize>,
+    machine_call_queue: BTreeSet<(&'queue QueueItem<'ast, T>, usize)>,
+    prover_function_queue: VecDeque<usize>,
+    occurrences: Rc<HashMap<Variable, Vec<usize>>>,
 }
 
-impl<'a, T: FieldElement> IdentityQueue<'a, T> {
+impl<'ast, 'queue, T: FieldElement> IdentityQueue<'ast, 'queue, T> {
     pub fn new(
-        fixed_data: &'a FixedData<'a, T>,
-        items: impl IntoIterator<Item = QueueItem<'a, T>>,
+        fixed_data: &'ast FixedData<'ast, T>,
+        items: &'queue Vec<QueueItem<'ast, T>>,
     ) -> Self {
-        let queue: BTreeSet<_> = items.into_iter().collect();
         let mut references = ReferencesComputer::new(fixed_data);
         let occurrences = Rc::new(
-            queue
+            items
                 .iter()
-                .flat_map(|item| {
+                .enumerate()
+                .flat_map(|(id, item)| {
                     references
                         .references(item)
                         .iter()
-                        .map(|v| (v.clone(), item.clone()))
+                        .map(|v| (v.clone(), id))
                         .collect_vec()
                 })
                 .into_group_map(),
         );
-        Self { queue, occurrences }
+        Self {
+            items,
+            in_queue: vec![true; items.len()],
+            identity_queue: (0..items.len())
+                .filter(|id| is_polynomial_identity_or_assignment(&items[*id]))
+                .collect(),
+            machine_call_queue: (0..items.len())
+                .filter_map(|id| is_submachine_call(&items[id]).then_some((&items[id], id)))
+                .collect(),
+            prover_function_queue: (0..items.len())
+                .filter(|id| is_prover_function(&items[*id]))
+                .collect(),
+            occurrences,
+        }
     }
 
     /// Returns the next identity to be processed and its row and
     /// removes it from the queue.
-    pub fn next(&mut self) -> Option<QueueItem<'a, T>> {
-        self.queue.pop_first()
+    pub fn next(&mut self) -> Option<&'queue QueueItem<'ast, T>> {
+        self.identity_queue
+            .pop_front()
+            .map(|id| {
+                self.in_queue[id] = false;
+                &self.items[id]
+            })
+            .or_else(|| {
+                self.machine_call_queue.pop_first().map(|(identity, id)| {
+                    self.in_queue[id] = false;
+                    identity
+                })
+            })
+            .or_else(|| {
+                self.prover_function_queue.pop_front().map(|id| {
+                    self.in_queue[id] = false;
+                    &self.items[id]
+                })
```

Comment:
As discussed, this means that machine calls might be processed out-of-order (because a prover function might be needed to make progress), but we'll deal with this later if needed (e.g. by re-ordering machine calls after all effects are computed).